bundles/framework/announcements/tool/AnnouncementsTool.js

```diff
@@ -0,0 +1,367 @@
+Oskari.clazz.define('Oskari.framework.announcements.tool.AnnouncementsTool',
+    function () {
+    }, {
+        index: 8,
+        pluginName: 'AnnouncementsPlugin',
+        allowedLocations: [],
+        allowedSiblings: [],
+        data: [],
+        selectedAnnouncements: [],
+        annTitles: [],
+        announcements: {},
+
+        isAnnouncementsDialogOpen: false,
+
+        groupedSiblings: true,
+
+        templates: {
+            announcements: jQuery('<div id="publisher-layout-announcements" class="tool-options">' + '<div id="publisher-layout-announcementsSelector">' + '<input type="text" name="publisher-announcements" id="publisher-announcements" disabled />' + '<button id="publisher-announcements-button"></button>' + '</div>' + '</div>'),
+            announcementsPopup: jQuery('<div id="publisher-announcements-popup">' + '<div id="publisher-announcements-inputs"></div>' + '</div>'),
+            inputCheckbox: jQuery('<div><input type="checkbox" /><label></label></div>')
+        },
+
+        eventHandlers: {
+            'Publisher2.ToolEnabledChangedEvent': function (event) {
+                var me = this;
+                var tool = event.getTool();
+                if (tool.getTool().id === me.getTool().id && tool.isStarted() && this.selectedAnnouncements) {
+                    me._sendAnnouncementsChangedEvent(this.selectedAnnouncements);
+                }
+            }
+        },
+
+        init: function (data) {
+            var me = this;
+            me.data = data;
+            me.selectedAnnouncements = [];
+            me.annTitles = [];
+
+            if (data.configuration && data.configuration.mapfull && data.configuration.mapfull.conf && data.configuration.mapfull.conf.plugins) {
+                _.each(data.configuration.mapfull.conf.plugins, function (plugin) {
+                    if (me.getTool().id === plugin.id) {
+                        me.setEnabled(true);
+                    }
+                });
+            }
+
+            for (var p in me.eventHandlers) {
+                if (me.eventHandlers.hasOwnProperty(p)) {
+                    me.__sandbox.registerForEventByName(me, p);
+                }
+            }
+            const toolPluginMapfullConf = this._getToolPluginMapfullConf();
+            if (toolPluginMapfullConf != null) {
+                me._sendAnnouncementsChangedEvent(toolPluginMapfullConf.config.announcements);
+                toolPluginMapfullConf.config.announcements.forEach(announcement => {
+                    me.annTitles.push(announcement.title);
+                    me.selectedAnnouncements.push(announcement);
+                });
+            }
+            this.fetchAnnouncements();
+            jQuery('div.basic_publisher').find('input[name=publisher-announcements]').val(me.annTitles.toString()).attr('announcement-name', me.annTitles.toString());
```

Comment:
Whatever this does, find another way of doing it.

Reply:
I heard this comment caused some grief so I'll try to rephrase: Can you tell me what this line does and why is it done. The problem I have with it is that it tries to find a single input from the whole publisher panel and set a string value and attribute as string but as far as I can see the variables are lists so I have no idea why this is done. Can you walk me through it?

Reply:
The input field is bit misleading, but basically we just get the (not editable) input field and set the selected announcements’ titles there. There could be a better way to do this, but it’s just implemented in the same style as in the ”styling” tool if I remember correctly.

Reply:
But isn't this already done in getExtraOptions() where it makes more sense? And what happens with fetchAnnouncements() above? It's just fired to load announcements to a variable which will complete at future time but nothing is triggered from them being loaded.

Reply:
I'm not sure if I understand what it is that is the expected behavior for this whole functionality. Since copies of announcements are saved to the plugin config and they are restored as choices here. But then the "actual" announcements are loaded as well but nothing seems to handle that really. So should the extra options list the currently available announcements in the instance. Or should it list the previously selected announcements. And what happens if the previously selected announcements are not longer part of the instances announcements? That's what doesn't make sense to me on this functionality as a whole.

Can you explain what the intended use is so I can make better evaluations for these changes. Making copies of announcements to the embedded maps is bad practice since you lose all control over the announcements. To me it would make more sense to save the announcement ids and fetch them by id on the embedded map. Then if they have expired -> no announcement is shown. And if they are edited the edited version is shown. As it is, it seems like a weird CMS to provide some popup notice on an embedded map BUT it requires both the announcements functionality and admin permissions for the user that is creating the embedded map to make it work in any sensible way. However since the publisher stuff for the announcements is shown for all users it feels a bit broken to me.